File: ts/src/cost_goal_audit.ts
import { existsSync, mkdirSync, readFileSync, writeFileSync } from "node:fs";
import { dirname } from "node:path";

type Args = {
  policyComparison?: string;
  resourceReport?: string;
  canaryValidation?: string;
  output?: string;
  format: "json" | "markdown";
  minReductionPct: number;
  minObservations: number;
};

type AuditCheck = {
  name: string;
  passed: boolean;
  evidence?: string;
  reason?: string;
};

function parseArgs(argv: string[]): Args {
  const values = new Map<string, string>();
  for (let index = 0; index < argv.length; index += 2) {
    values.set(argv[index], argv[index + 1]);
  }
  return {
    policyComparison: values.get("--policy-comparison"),
    resourceReport: values.get("--resource-report"),
    canaryValidation: values.get("--canary-validation"),
    output: values.get("--output"),
    format: parseFormat(values.get("--format") ?? "markdown"),
    minReductionPct: Number.parseFloat(values.get("--min-reduction-pct") ?? "20"),
    minObservations: Number.parseInt(values.get("--min-observations") ?? "1", 10)
  };
}

function parseFormat(value: string): Args["format"] {
  if (value === "json" || value === "markdown") {
    return value;
  }
  throw new Error(`Unsupported format: ${value}`);
}

function policyComparisonCheck(args: Args): AuditCheck {
  if (!args.policyComparison) {
    return { name: "policy_projection", passed: false, reason: "--policy-comparison not provided" };
  }
  const parsed = readJson(args.policyComparison);
  const reduction = numberAt(parsed, "total_reduction_pct");
  if (reduction === undefined) {
    return { name: "policy_projection", passed: false, reason: "total_reduction_pct missing" };
  }
  return {
    name: "policy_projection",
    passed: reduction >= args.minReductionPct,
    evidence: `${reduction.toFixed(1)}% projected reduction`,
    ...(reduction >= args.minReductionPct ? {} : { reason: `below ${args.minReductionPct.toFixed(1)}%` })
  };
}

function resourceObservationCheck(args: Args): AuditCheck {
  if (!args.resourceReport) {
    return { name: "resource_observations", passed: false, reason: "--resource-report not provided" };
  }
  const parsed = readJson(args.resourceReport);
  const count = numberAt(parsed, "observation_count");
  if (count === undefined) {
    return { name: "resource_observations", passed: false, reason: "observation_count missing" };
  }
  return {
    name: "resource_observations",
    passed: count >= args.minObservations,
    evidence: `${count} observations`,
    ...(count >= args.minObservations ? {} : { reason: `below ${args.minObservations}` })
  };
}

function canaryValidationCheck(args: Args): AuditCheck {
  if (!args.canaryValidation) {
    return {
      name: "remote_canary_validation",
      passed: false,
      reason: "--canary-validation not provided"
    };
  }
  if (!existsSync(args.canaryValidation)) {
    return {
      name: "remote_canary_validation",
      passed: false,
      reason: `canary validation file not found: ${args.canaryValidation}`
    };
  }
  const parsed = readJson(args.canaryValidation);
  const passed = booleanAt(parsed, "passed");
  const reduction = numberAt(parsed, "total_reduction_pct");
  if (passed !== true) {
    return {
      name: "remote_canary_validation",
      passed: false,
      evidence: reduction === undefined ? undefined : `${reduction.toFixed(1)}% actual reduction`,
      reason: "canary validator did not pass"
    };
  }
  if (reduction === undefined || reduction < args.minReductionPct) {
    return {
      name: "remote_canary_validation",
      passed: false,
      evidence: reduction === undefined ? undefined : `${reduction.toFixed(1)}% actual reduction`,
      reason: `actual reduction below ${args.minReductionPct.toFixed(1)}%`
    };
  }
  return {
    name: "remote_canary_validation",
    passed: true,
    evidence: `${reduction.toFixed(1)}% actual reduction`
  };
}

function readJson(path: string): unknown {
  if (!existsSync(path)) {
    throw new Error(`Input not found: ${path}`);
  }
  return JSON.parse(readFileSync(path, "utf8")) as unknown;
}

function numberAt(value: unknown, key: string): number | undefined {
  if (!isObject(value)) {
    return undefined;
  }
  const item = value[key];
  return typeof item === "number" && Number.isFinite(item) ? item : undefined;
}

function booleanAt(value: unknown, key: string): boolean | undefined {
  if (!isObject(value)) {
    return undefined;
  }
  const item = value[key];
  return typeof item === "boolean" ? item : undefined;
}

function isObject(value: unknown): value is Record<string, unknown> {
  return typeof value === "object" && value !== null;
}

function markdownReport(checks: AuditCheck[], args: Args): string {
  return [
    "# Cost Reduction Goal Audit",
    "",
    `Generated: ${new Date().toISOString()}`,
    "",
    `Minimum reduction: ${args.minReductionPct.toFixed(1)}%`,
    `Minimum observations: ${args.minObservations}`,
    "",
    "check | status | evidence | reason",
    "--- | --- | --- | ---",
    ...checks.map((check) =>
      [check.name, check.passed ? "pass" : "fail", check.evidence ?? "-", check.reason ?? "-"].join(" | ")
    ),
    "",
    `Overall: ${checks.every((check) => check.passed) ? "pass" : "fail"}`,
    ""
  ].join("\n");
}

function writeOutput(path: string | undefined, content: string): void {
  if (!path) {
    console.log(content);
    return;
  }
  mkdirSync(dirname(path), { recursive: true });
  writeFileSync(path, content);
  console.log(`wrote ${path}`);
}

async function main(): Promise<void> {
  const args = parseArgs(Bun.argv.slice(2));
  const checks = [policyComparisonCheck(args), resourceObservationCheck(args), canaryValidationCheck(args)];
  const passed = checks.every((check) => check.passed);
  const payload = {
    schema_version: 1,
    generated_at: new Date().toISOString(),
    min_reduction_pct: args.minReductionPct,
    min_observations: args.minObservations,
    passed,
    checks
  };
  const content = args.format === "json" ? `${JSON.stringify(payload, null, 2)}\n` : markdownReport(checks, args);
  writeOutput(args.output, content);
  if (!passed) {
    process.exitCode = 1;
  }
}

if (import.meta.main) {
  await main();
}

File: py/code_sandbox_bench/cost_goal_audit.py
import argparse
import json
import math
from datetime import datetime, timezone
from pathlib import Path
from typing import Any


def iso_now() -> str:
    return datetime.now(timezone.utc).isoformat().replace("+00:00", "Z")


def parse_args() -> argparse.Namespace:
    parser = argparse.ArgumentParser()
    parser.add_argument("--policy-comparison", type=Path)
    parser.add_argument("--resource-report", type=Path)
    parser.add_argument("--canary-validation", type=Path)
    parser.add_argument("--output", type=Path)
    parser.add_argument("--format", choices=["json", "markdown"], default="markdown")
    parser.add_argument("--min-reduction-pct", type=float, default=20.0)
    parser.add_argument("--min-observations", type=int, default=1)
    return parser.parse_args()


def policy_comparison_check(args: argparse.Namespace) -> dict[str, Any]:
    if not args.policy_comparison:
        return {"name": "policy_projection", "passed": False, "reason": "--policy-comparison not provided"}
    parsed = read_json(args.policy_comparison)
    reduction = number_at(parsed, "total_reduction_pct")
    if reduction is None:
        return {"name": "policy_projection", "passed": False, "reason": "total_reduction_pct missing"}
    return {
        "name": "policy_projection",
        "passed": reduction >= args.min_reduction_pct,
        "evidence": f"{reduction:.1f}% projected reduction",
        **({} if reduction >= args.min_reduction_pct else {"reason": f"below {args.min_reduction_pct:.1f}%"}),
    }


def resource_observation_check(args: argparse.Namespace) -> dict[str, Any]:
    if not args.resource_report:
        return {"name": "resource_observations", "passed": False, "reason": "--resource-report not provided"}
    parsed = read_json(args.resource_report)
    count = number_at(parsed, "observation_count")
    if count is None:
        return {"name": "resource_observations", "passed": False, "reason": "observation_count missing"}
    return {
        "name": "resource_observations",
        "passed": count >= args.min_observations,
        "evidence": f"{int(count)} observations",
        **({} if count >= args.min_observations else {"reason": f"below {args.min_observations}"}),
    }


def canary_validation_check(args: argparse.Namespace) -> dict[str, Any]:
    if not args.canary_validation:
        return {"name": "remote_canary_validation", "passed": False, "reason": "--canary-validation not provided"}
    if not args.canary_validation.exists():
        return {"name": "remote_canary_validation", "passed": False, "reason": f"canary validation file not found: {args.canary_validation}"}
    parsed = read_json(args.canary_validation)
    passed = parsed.get("passed") if isinstance(parsed, dict) else None
    reduction = number_at(parsed, "total_reduction_pct")
    if passed is not True:
        return {
            "name": "remote_canary_validation",
            "passed": False,
            **({"evidence": f"{reduction:.1f}% actual reduction"} if reduction is not None else {}),
            "reason": "canary validator did not pass",
        }
    if reduction is None or reduction < args.min_reduction_pct:
        return {
            "name": "remote_canary_validation",
            "passed": False,
            **({"evidence": f"{reduction:.1f}% actual reduction"} if reduction is not None else {}),
            "reason": f"actual reduction below {args.min_reduction_pct:.1f}%",
        }
    return {"name": "remote_canary_validation", "passed": True, "evidence": f"{reduction:.1f}% actual reduction"}


def markdown_report(checks: list[dict[str, Any]], args: argparse.Namespace) -> str:
    return "\n".join(
        [
            "# Cost Reduction Goal Audit",
            "",
            f"Generated: {iso_now()}",
            "",
            f"Minimum reduction: {args.min_reduction_pct:.1f}%",
            f"Minimum observations: {args.min_observations}",
            "",
            "check | status | evidence | reason",
            "--- | --- | --- | ---",
            *[
                " | ".join([check["name"], "pass" if check["passed"] else "fail", check.get("evidence") or "-", check.get("reason") or "-"])
                for check in checks
            ],
            "",
            f"Overall: {'pass' if all(check['passed'] for check in checks) else 'fail'}",
            "",
        ]
    )


def read_json(path: Path) -> dict[str, Any]:
    if not path.exists():
        raise FileNotFoundError(f"Input not found: {path}")
    parsed = json.loads(path.read_text(encoding="utf-8"))
    if not isinstance(parsed, dict):
        raise ValueError(f"Expected JSON object in {path}")
    return parsed


def number_at(value: Any, key: str) -> float | None:
    item = value.get(key) if isinstance(value, dict) else None
    return float(item) if isinstance(item, (int, float)) and math.isfinite(item) else None


def write_output(path: Path | None, content: str) -> None:
    if path is None:
        print(content)
        return
    path.parent.mkdir(parents=True, exist_ok=True)
    path.write_text(content, encoding="utf-8")
    print(f"wrote {path}")


def main() -> None:
    args = parse_args()
    checks = [policy_comparison_check(args), resource_observation_check(args), canary_validation_check(args)]
    payload = {
        "schema_version": 1,
        "generated_at": iso_now(),
        "min_reduction_pct": args.min_reduction_pct,
        "min_observations": args.min_observations,
        "passed": all(check["passed"] for check in checks),
        "checks": checks,
    }
    content = json.dumps(payload, indent=2) + "\n" if args.format == "json" else markdown_report(checks, args)
    write_output(args.output, content)
    if not payload["passed"]:
        raise SystemExit(1)


if __name__ == "__main__":
    main()
